Simplify Windows linking in ExtUtils::CBuilder

Leont opened the pull request:
This is a follow-up on @zmughal's 4371f648bf908d4ac28660f37f560b0f8fbd0488. Now that it uses `--enable-auto-image-base` the previous and complicated `dlltool; ld; dlltool; ld` dance is no longer necessary.

This makes a previously dysfunctional test succeed, and hopefully also fixes build issues as seen in https://github.com/Leont/magic-check/issues/1

@ambs

Leont committed:
Simplify Windows linking in ExtUtils::CBuilder

zmughal commented on dist/ExtUtils-CBuilder/lib/ExtUtils/CBuilder/Platform/Windows/GCC.pm:
```diff
@@ -95,8 +64,6 @@ sub format_linker_cmd {
     $spec{explib}             ,
```
The `.exp` file is not created anymore. Testing this change in a VM:

```
# :!gmake test

"C:\Strawberry\perl\bin\perl.exe" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib\lib', 'blib\arch')" t/*.t
t/00-have-compiler.t .. ok
t/01-basic.t .......... g++: error: compilet.exp: No such file or directory
t/01-basic.t .......... skipped: no compiler available for testing
t/02-link.t ........... skipped: link_executable() is not implemented yet on MSWin32
t/03-cplusplus.t ...... g++: error: compilet.exp: No such file or directory
t/03-cplusplus.t ...... skipped: no compiler available for testing
t/04-base.t ........... ok
All tests successful.
Files=5, Tests=72,  3 wallclock secs ( 0.17 usr +  0.02 sys =  0.19 CPU)
Result: PASS


# :!perl -v

This is perl 5, version 32, subversion 1 (v5.32.1) built for MSWin32-x64-multi-thread

Copyright 1987-2021, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.


# :!gcc -v

Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=C:/Strawberry/c/bin/../libexec/gcc/x86_64-w64-mingw32/8.3.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../../../src/gcc-8.3.0/configure --host=x86_64-w64-mingw32 --build=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --prefix=/mingw64 --enable-shared --enable-static --disable-multilib --enable-languages=c,c++,fortran,lto --enable-libstdcxx-time=yes --enable-threads=posix --enable-libgomp --enable-libatomic --enable-lto --enable-graphite --enable-checking=release --enable-fully-dynamic-string --enable-versio
n-specific-runtime-libs --enable-libstdcxx-filesystem-ts=yes --disable-libstdcxx-pch --disable-libstdcxx-debug --disable-bootstrap --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --with-gnu-as --with-gnu-ld --with-arch=nocona --with-tune=core2 --with-libiconv --with-system-zlib --with-gmp=/opt/build/prerequisites/x86_64-w64-mingw32-static --with-mpfr=/opt/build/prerequisites/x86_64
-w64-mingw32-static --with-mpc=/opt/build/prerequisites/x86_64-w64-mingw32-static --with-isl=/opt/build/prerequisites/x86_64-w64-mingw32-static --with-pkgversion='x86_64-posix-seh, Built by strawberryperl.com project' --with-bugurl=https://sourceforge.net/projects/mingw-w64 CFLAGS='-O2 -pipe -fno-ident -I/opt/build/x86_64-830-posix-seh-rt_v6/mingw64/opt/include -I/opt/build/prerequisites/x86_64-zlib-static/include -I/opt/b
uild/prerequisites/x86_64-w64-mingw32-static/include' CXXFLAGS='-O2 -pipe -fno-ident -I/opt/build/x86_64-830-posix-seh-rt_v6/mingw64/opt/include -I/opt/build/prerequisites/x86_64-zlib-static/include -I/opt/build/prerequisites/x86_64-w64-mingw32-static/include' CPPFLAGS=' -I/opt/build/x86_64-830-posix-seh-rt_v6/mingw64/opt/include -I/opt/build/prerequisites/x86_64-zlib-static/include -I/opt/build/prerequisites/x86_64-w64-mi
ngw32-static/include' LDFLAGS='-pipe -fno-ident -L/opt/build/x86_64-830-posix-seh-rt_v6/mingw64/opt/lib -L/opt/build/prerequisites/x86_64-zlib-static/lib -L/opt/build/prerequisites/x86_64-w64-mingw32-static/lib ' LD_FOR_TARGET=/opt/build/x86_64-830-posix-seh-rt_v6/mingw64/bin/ld.exe
Thread model: posix
gcc version 8.3.0 (x86_64-posix-seh, Built by strawberryperl.com project)
```

Removing the line fixes the test:

```patch
-    $spec{explib}             ,
```

```
# :!prove -bvr t\01-basic.t t\03-cplusplus.t

t\01-basic.t ......
1..12
ok 1 - created EU::CB object
ok 2 - have_compiler
ok 3 - source file 't\basict.c' created
ok 4
gcc -c -DWIN32 -DWIN64 -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2 -I"C:\STRAWB~1\perl\lib\CORE" -I"C:\STRAWB~1\c\include" -o "t\basict.o" "t\basict.c"
ok 5
ok 6
Generating script 't\basict.lds'
g++ -o "t\basict.xs.dll" -Wl,--enable-auto-image-base -mdll -s -L"C:\STRAWB~1\perl\lib\CORE" -L"C:\STRAWB~1\c\lib" "t\basict.lds"
ok 7
ok 8 # skip MSWindows
ok 9 # skip MSWindows
ok 10 # skip MSWindows
ok 11
ok 12
ok
t\03-cplusplus.t ..
1..7
ok 1 - created EU::CB object
ok 2 - have_cplusplus
ok 3 - source file 't\cplust.cc' created
ok 4
gcc -c -DWIN32 -DWIN64 -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2 -I"C:\STRAWB~1\perl\lib\CORE" -I"C:\STRAWB~1\c\include" -o "t\cplust.o" "t\cplust.cc"
ok 5
ok 6
Generating script 't\cplust.lds'
g++ -o "t\cplust.xs.dll" -Wl,--enable-auto-image-base -mdll -s -L"C:\STRAWB~1\perl\lib\CORE" -L"C:\STRAWB~1\c\lib" "t\cplust.lds"
ok 7
ok
All tests successful.
Files=2, Tests=19,  2 wallclock secs ( 0.05 usr +  0.00 sys =  0.05 CPU)
Result: PASS
```

zmughal replied:
Just to catch this earlier, perhaps the test should make sure that the simplest possible compilation works instead of skipping with `skipped: no compiler available for testing`? Though that would be kind of annoying as that is different from what this module is made to do.

zmughal reviewed (changes requested):
Looks reasonable to me.

Though I am seeing the following failure when the downstream `ExtUtils::CppGuess`
is installed:


```
# :!cpanm --test-only --verbose ExtUtils::CppGuess

--> Working on ExtUtils::CppGuess
Fetching http://www.cpan.org/authors/id/E/ET/ETJ/ExtUtils-CppGuess-0.26.tar.gz ... OK
Unpacking ExtUtils-CppGuess-0.26.tar.gz
...
t/010_module_build.t ... # Module::Build version: 0.4231
# ExtUtils::CBuilder version: 0.280240

#   Failed test 'build with Module::Build'
#   at t/010_module_build.t line 19.
t/010_module_build.t ... 1/? # Build.PL output
# ========================================
# Created MYMETA.yml and MYMETA.json
# Creating new 'Build' script for 'CppGuessTest' version '0.01'
# ========================================
# Build output
# ========================================
# Building CppGuessTest
# g++ -c -DWIN32 -DWIN64 -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -xc++ -DINCLUDE_DOT=0 -s -O2 "-DVERSION=\"0.01\"" "-DXS_VERSION=\"0.01\"" -I"C:\STRAWB~1\perl\lib\CORE" -I"C:\STRAWB~1\c\include" -o "lib\CppGuessTest.o" "lib\CppGuessTest.c"
# ExtUtils::Mkbootstrap::Mkbootstrap('blib\arch\auto\CppGuessTest\CppGuessTest.bs')
# Generating script 'lib\CppGuessTest.lds'
# g++ -o "blib\arch\auto\CppGuessTest\CppGuessTest.xs.dll" -Wl,--enable-auto-image-base -mdll -s -L"C:\STRAWB~1\perl\lib\CORE" -L"C:\STRAWB~1\c\lib" "lib\CppGuessTest.lds" -lstdc++
# ========================================
# Build test output
# ========================================
#
# #   Failed test 'use CppGuessTest;'
# #   at t\001_load.t line 4.
# #     Tried to use 'CppGuessTest'.
# #     Error:  Can't find 'boot_CppGuessTest' symbol in C:\Users\vagrant\.cpanm\work\1694307910.1956\ExtUtils-CppGuess-0.26\t\module_build\blib\arch/auto/CppGuessTest/CppGuessTest.xs.dll
# #  at t\001_load.t line 4.
# # Compilation failed in require at t\001_load.t line 4.
# # BEGIN failed--compilation aborted at t\001_load.t line 4.
# # Looks like you failed 1 test of 1.
# t\001_load.t ..
# Dubious, test returned 1 (wstat 256, 0x100)
# Failed 1/1 subtests
# Can't find 'boot_CppGuessTest' symbol in C:\Users\vagrant\.cpanm\work\1694307910.1956\ExtUtils-CppGuess-0.26\t\module_build\blib\arch/auto/CppGuessTest/CppGuessTest.xs.dll
#  at t\002_base.t line 4.
# Compilation failed in require at t\002_base.t line 4.
# BEGIN failed--compilation aborted at t\002_base.t line 4.
# t\002_base.t ..
# Dubious, test returned 2 (wstat 512, 0x200)
# Failed 2/2 subtests
#
# Test Summary Report
# -------------------
# t\001_load.t (Wstat: 256 Tests: 1 Failed: 1)
#   Failed test:  1
#   Non-zero exit status: 1
# t\002_base.t (Wstat: 512 Tests: 0 Failed: 0)
#   Non-zero exit status: 2
#   Parse errors: Bad plan.  You planned 2 tests but ran 0.
# Files=2, Tests=1,  0 wallclock secs ( 0.03 usr +  0.02 sys =  0.05 CPU)
# Result: FAIL
# Failed 2/2 test programs. 1/1 subtests failed.
# ========================================
# Looks like you failed 1 test of 1.
t/010_module_build.t ... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
t/011_makemaker.t ...... ok

Test Summary Report
-------------------
t/010_module_build.t (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=7, Tests=20, 14 wallclock secs ( 0.06 usr +  0.03 sys =  0.09 CPU)
Result: FAIL
Failed 1/7 test programs. 1/20 subtests failed.
gmake: *** [Makefile:885: test_dynamic] Error 255
FAIL
! Testing ExtUtils::CppGuess failed. See C:\Users\vagrant\.cpanm\work\1694307910.1956\build.log for details. Retry with --force to force install it.
```

This failure is not present with `AMBS/ExtUtils-CBuilder-0.280236.tar.gz`:

```
t/001_load.t ........... ok
t/002_icpp.t ........... ok
t/003_standard_flag.t .. 1/? g++: error: unrecognized command line option '-std=c++unreal'; did you mean '-std=c++2a'?
t/003_standard_flag.t .. ok
t/004_env.t ............ 1/? # EUMM env:{
#   'CC' => 'czz',
#   'CCFLAGS' => ' -DWIN32 -DWIN64 -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -xc++',
#   'dynamic_lib' => {
#     'OTHERLDFLAGS' => ' -lstdc++'
#   }
# }
t/004_env.t ............ ok
t/010_module_build.t ... # Module::Build version: 0.4231
# ExtUtils::CBuilder version: 0.280236
t/010_module_build.t ... ok
t/011_makemaker.t ...... ok
All tests successful.
Files=7, Tests=20, 15 wallclock secs ( 0.12 usr +  0.06 sys =  0.19 CPU)
Result: PASS
OK
Successfully tested ExtUtils-CppGuess-0.26
```

Leont commented:
> Removing the line fixes the test:

I did just that, does that also fix the cppguess issue?

zmughal commented:
> > Removing the line fixes the test:
> 
> I did just that, does that also fix the cppguess issue?

Sorry, I just realised that I was not clear in my review comment: the above issue with `ExtUtils::CppGuess` was with the change I recommended (the compilation would fail otherwise). So `ExtUtils::CppGuess` still needs fixing.

tonycoz commented:
I'm confused as to whether this is ready for review.

It looks sane enough to me.

Leont commented:
> I'm confused as to whether this is ready for review.

> It looks sane enough to me.

It was, but @zmughal noticed a BBC that I hope I have fixed now.

zmughal commented:
With https://github.com/Perl/perl5/commit/b60f45b9f1f0d4d4dc948dd8c3223eec3ca65d55 : 

```
:!prove -bvr t\03-cplusplus.t


t\03-cplusplus.t .. C:/Strawberry/c/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: compilet.exp:fake:(.edata+0x3c): undefined reference to `boot_compilet'
collect2.exe: error: ld returned 1 exit status

1..0 # SKIP no compiler available for testing
skipped: no compiler available for testing
Files=1, Tests=0,  1 wallclock secs ( 0.09 usr +  0.00 sys =  0.09 CPU)
Result: NOTESTS
```


```
:!cpanm --test-only --verbose ExtUtils::CppGuess

Searching ExtUtils::CppGuess () on cpanmetadb ...
--> Working on ExtUtils::CppGuess
Fetching http://www.cpan.org/authors/id/E/ET/ETJ/ExtUtils-CppGuess-0.26.tar.gz ... OK
Unpacking ExtUtils-CppGuess-0.26.tar.gz
Entering ExtUtils-CppGuess-0.26
...
t/010_module_build.t ... # Module::Build version: 0.4231
# ExtUtils::CBuilder version: 0.280240

t/010_module_build.t ... 1/? #   Failed test 'build with Module::Build'
#   at t/010_module_build.t line 19.
# Build.PL output
# ========================================
# C:/Strawberry/c/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: compilet.exp:fake:(.edata+0x3c): undefined reference to `boot_compilet'
# collect2.exe: error: ld returned 1 exit status
# Warning: ExtUtils::CBuilder not installed or no compiler detected
# Proceeding with configuration, but compilation may fail during Build
#
# Created MYMETA.yml and MYMETA.json
# Creating new 'Build' script for 'CppGuessTest' version '0.01'
# ========================================
# Build output
# ========================================
# Error: no compiler detected to compile 'lib\CppGuessTest.c'.  Aborting
# Building CppGuessTest
# ========================================
# Looks like you failed 1 test of 1.
t/010_module_build.t ... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
t/011_makemaker.t ...... ok

Test Summary Report
-------------------
t/010_module_build.t (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=7, Tests=20, 68 wallclock secs ( 0.12 usr +  0.01 sys =  0.14 CPU)
Result: FAIL
Failed 1/7 test programs. 1/20 subtests failed.
gmake: *** [Makefile:885: test_dynamic] Error 255
FAIL
! Testing ExtUtils::CppGuess failed. See C:\Users\vagrant\.cpanm\work\1694646750.5196\build.log for details. Retry with --force to force install it.
```

zmughal commented:
Re https://github.com/Perl/perl5/commit/b60f45b9f1f0d4d4dc948dd8c3223eec3ca65d55 again, this is just to show that there is a work around if you add: `'-Wl,--export-all-symbols'`. So the symbols are there and EU::CppGuess's have been working via `MakeMaker` (which also dropped `dlltool` and uses `--enable-auto-image-base`).

These are the commands that `MakeMaker` uses (no linker script):

```
"C:\Strawberry\perl\bin\perl.exe" -MExtUtils::Command -e rm_rf -- blib\arch\auto\CppGuessTest\CppGuessTest.bs
"C:\Strawberry\perl\bin\perl.exe" -MExtUtils::Command::MM -e cp_nonempty -- CppGuessTest.bs blib\arch\auto\CppGuessTest\CppGuessTest.bs 644
"C:\Strawberry\perl\bin\perl.exe" "C:\Strawberry\perl\lib\ExtUtils/xsubpp"  -typemap C:\STRAWB~1\perl\lib\ExtUtils\typemap -typemap C:\b\extutils-cppguess\t\makemaker\typemap  CppGuessTest.xs > CppGuessTest.xsc
"C:\Strawberry\perl\bin\perl.exe" -MExtUtils::Command -e mv -- CppGuessTest.xsc CppGuessTest.c
g++ -c   -DWIN32 -DWIN64 -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -xc++ -DINCLUDE_DOT=0 -s -O2   -DVERSION=\"0.01\" -DXS_VERSION=\"0.01\"  "-IC:\STRAWB~1\perl\lib\CORE"   CppGuessTest.c
g++ CppGuessTest.def -o blib\arch\auto\CppGuessTest\CppGuessTest.xs.dll -mdll -s -L"C:\STRAWB~1\perl\lib\CORE" -L"C:\STRAWB~1\c\lib" CppGuessTest.o -lstdc++  "C:\STRAWB~1\perl\lib\CORE\libperl532.a" -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32 -Wl,--enable-auto-image-base
"C:\Strawberry\perl\bin\perl.exe" -MExtUtils::Command -e chmod -- 755 blib\arch\auto\CppGuessTest\CppGuessTest.xs.dll

```

zmughal committed:
Make the C++ test export the expected symbol

Use `namespace` instead of `class` and mark as `extern "C"`.

zmughal committed:
Remove dlltool and use .def file directly

zmughal committed:
Make sure linking works even when CC is set to C++ compiler

This can happen when using a module such as `ExtUtils::CppGuess` which
sets `CC` and `CCFLAGS` to C++ compiler specific values. When combined
with `Module::Build` which calls `ExtUtils::CBuilder::Base::have_compiler`
with those values, the branch that checks for a C compiler is run with
those C++ compiler specific values which causes a linker error.

zmughal commented:
I am satisfied with this change as it does the same thing as `ExtUtils::MakeMaker` now. I see no issues in the test output:

```sh
gh -R Perl/perl5 run view 6195173479 --json jobs -q '.jobs.[].databaseId' \
  |  xargs -I{} sh -c 'gh -R Perl/perl5 run view --job=$0 --log | grep -C5 "CBuilder.*\.t" ' {} \
  | $PAGER
```
